xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp

```diff
@@ -1151,6 +1151,8 @@ void CAESinkPULSE::Drain()
 
   pa_threaded_mainloop_lock(m_MainLoop);
   WaitForOperation(pa_stream_drain(m_Stream, NULL, NULL), m_MainLoop, "Drain");
+  WaitForOperation(pa_stream_cork(m_Stream, 1, NULL, NULL), m_MainLoop, "Pause");
```

Comment:
you call the API function internally. did you verify that internal contract did not break?

Reply:
I made a runtime test and checked the API documentation:
https://freedesktop.org/software/pulseaudio/doxygen/stream_8h.html#a14e698233ac2d246646651955ab0ec7b 

Did you spot something i break?

Reply:
Or do you mean I should use: pa_stream_is_corked()  and only cork if not corked?

Reply:
```
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] protocol-native.c: Implicit drain of 'kodi audio stream'
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 3392 bytes ago (6144 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 3392 bytes ago (6080 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 3392 bytes ago (5984 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 3392 bytes ago (4032 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 3392 bytes ago (3904 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 9440 bytes ago (96 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink.c: alsa_output.pci-0000_00_1f.3.analog-stereo: Found underrun 9440 bytes ago (32 bytes ahead in playback buffer)
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] protocol-native.c: Drain acknowledged of 'kodi audio stream'
Nov 25 18:38:27 L380 pulseaudio[18207]: D: [alsa-sink-ALC257 Analog] sink-input.c: Requesting rewind due to corking
```

From pulse's log - it looks okay. It runs down and then pauses.

Reply:
I did not look into the details. Just saying that you used the API function Drain() internally too. If you change behaviour for the outside world, just make sure that it does not break the internal stuff.

Reply:
My reasoning was:
- Drain() Play out all buffers (snd_pcm_drain)
- Cork() afterwards, stop the Sink, so that we can refill it (altternative to snd_pcm_prepare)

AddPacket is checking if sink is currently corked and uncorks it when adding first package.

Sounds sane?

Reply:
> Sounds sane?

Yes, this is the contract for the external world but you call Drain() from Deinitialize(), too. Why would you pause a sink you about to close? Apart from that calling Drain() from Deinitialize makes no sense to me at all. When AE calls Deiinitialize it wants exactly that, not waiting for an unspecified time for this function to complete. That can be 4 seconds and more for network sinks.

Reply:
Yes. You are right. Should I flush on deinitialize instead?

https://freedesktop.org/software/pulseaudio/doxygen/stream_8h.html#ac9c524eb275d258b00c553409682409c

Reply:
Looking at Alsa flush makes perfect sense. In ALSA we use drop. Will add a commit.